cli/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,70 @@
+package healthcheck
+
+type CheckStatus string
+
+const (
+	CheckOk     = CheckStatus("OK")
+	CheckFailed = CheckStatus("FAIL")
+	CheckError  = CheckStatus("ERROR")
+)
+
+type CheckResult struct {
+	SubsystemName    string
+	CheckDescription string
+	Status           CheckStatus
+	NextSteps        string
+}
+
+type Check struct {
+	Results       []CheckResult
+	OverallStatus CheckStatus
+}
+
+type StatusChecker interface {
+	SelfCheck() ([]CheckResult, error)
+}
+
+type CheckObserver func(result CheckResult)
+
+type HealthChecker struct {
+	subsystemsToCheck []StatusChecker
+}
+
+func (hC *HealthChecker) Add(subsystemChecker StatusChecker) {
+	hC.subsystemsToCheck = append(hC.subsystemsToCheck, subsystemChecker)
+}
+
+func (hC *HealthChecker) PerformCheck(observer CheckObserver) Check {
+	if observer == nil {
+		observer = func(_ CheckResult) {}
+	}
+
+	check := Check{
+		Results:       make([]CheckResult, 0),
+		OverallStatus: CheckOk,
+	}
+
+	for _, checker := range hC.subsystemsToCheck {
+		results, err := checker.SelfCheck()
```

Comment:
it's possible for `SelfCheck()` to return:
1) `[]{CheckOk}`, `nil`
2) `[]{CheckFailed}`, `nil`
3) `[]{CheckError}`, `nil`
4) `[]{CheckOk}`, `err`
5) `[]{CheckFailed}`, `err`
6) `[]{CheckError}`, `err`

... is it possible to tighten this up? what would it mean for the function to error but all results come back `CheckOk`?

since we're already setting `check.OverallStatus = CheckError` anytime we receive any CheckError, can we just consider an `err` returned by `SelfCheck()` as a real program failure, and exit?

or, we only consider checkers as OK/FAIL, and errors always cause program exit?

Reply:
I'm not sure what would be a good way out here. I've added 09c88537163fef42b8a6fc8e023178622599a0b1, do you think it gets closer to a better state?

Reply:
i think the continue added makes more sense, assuming `SelfCheck()` returning an error means that `results` should not be evaluated.

i think i'm getting caught up with individual checks returning ok/fail/err and `SelfCheck()` also returning ok/err, and then all of them feeding into `OverallStatus`.

not sure if this would work, but in my mind the simplest approach would be for individual checks to only return OK or FAIL, and for `SelfCheck()` to return OK or ERR. if we get an ERR from `SelfCheck()` we could either bail out immediately or carry on but communicate to the user that something went unexpectedly wrong.

fwiw this is all a bit of a yak shave at this point, i'd be fine leaving this as is, with or without https://github.com/runconduit/conduit/commit/09c88537163fef42b8a6fc8e023178622599a0b1.

Reply:
@siggy this seems to be too close to home with my current go question: is it ok to return an error and a mostly working object? Should people use that object? But besides the fun philosophical question I agree, let's see how this changes with new checks and at some point crystalise some norm around it.

Reply:
it occurred to me that with this `continue`, it's possible for `OverallStatus` to be `CheckError`, but without any sub-results to show for it. this could be confusing for the user, maybe just log the error or something?